src/platforms/evdev/libinput_device.cpp

```diff
@@ -358,12 +361,26 @@ mir::EventUPtr mie::LibInputDevice::convert_touch_frame(libinput_event_touch* to
         if (data.action == mir_touch_action_down)
             data.action = mir_touch_action_change;
 
-        if (data.action == mir_touch_action_up)
+        if (data.action == mir_touch_action_up || data.down_notified)
             it = last_seen_properties.erase(it);
```

Comment:
This (`|| data.down_notified`) looks wrong!

If (on a previous frame) we notified a "down" then we drop the contact data regardless of what the current action is. So if the current action is a `LIBINPUT_EVENT_TOUCH_MOTION`, then our internal state changes to "up" and we never pass on the release.